src/AutoMapper/MapperConfiguration.cs

```diff
@@ -66,25 +66,66 @@ public MapperConfiguration(Action<IMapperConfigurationExpression> configure, IEn
         {
             var key = new TypePair(typeof(TSource), typeof(TDestination));
             var mapRequest = new MapRequest(key, types);
-            return (Func<TSource, TDestination, ResolutionContext, TDestination>)GetMapperFunc(mapRequest);
+            var mapperFuncs = GetMapperFuncs(mapRequest);
+            return GetSubMap<TSource, TDestination>(mapperFuncs)?.Compile() ?? (Func<TSource, TDestination, ResolutionContext, TDestination>)mapperFuncs.Typed;
+        }
+
+        public Expression<Func<TSource, TDestination, ResolutionContext, TDestination>> GetMapperExpression<TSource, TDestination>(TypePair typePair)
+        {
+            var key = new TypePair(typeof(TSource), typeof(TDestination));
+            var mapRequest = new MapRequest(key, typePair);
+            var mapperFuncs = GetMapperFuncs(mapRequest);
+            return GetSubMap<TSource, TDestination>(mapperFuncs) ?? (Expression<Func<TSource, TDestination, ResolutionContext, TDestination>>)mapperFuncs.TypedExpression;
+        }
+
+        private Expression<Func<TSource, TDestination, ResolutionContext, TDestination>> GetSubMap<TSource, TDestination>(MapperFuncs mapperFuncs)
+        {
+            var mapExpression = mapperFuncs.TypedExpression;
+            if (mapExpression != null && (mapExpression.Parameters[0].Type != typeof(TSource) ||
+                mapExpression.Parameters[1].Type != typeof(TDestination)))
+            {
+                var requestedSourceParameter = Parameter(typeof(TSource), "src");
+                var requestedDestinationParameter = Parameter(typeof(TDestination), "dest");
+                var contextParameter = Parameter(typeof(ResolutionContext), "ctxt");
+                return (Expression<Func<TSource, TDestination, ResolutionContext, TDestination>>)
+                    Lambda(
+                        ToType(
+                            mapExpression.ReplaceParameters(
+                                ToType(requestedSourceParameter, mapExpression.Parameters[0].Type),
+                                ToType(requestedDestinationParameter, mapExpression.Parameters[1].Type),
+                                contextParameter),
+                            typeof(TDestination)),
+                        requestedSourceParameter, requestedDestinationParameter, contextParameter);
+            }
+            return null;
+        }
+
+        private MapperFuncs GetMapperFuncs(MapRequest mapRequest)
+        {
+            var mapperFunc = _mapPlanCache.GetOrAdd(mapRequest);
+            if (mapperFunc.TypedExpression == null)
+                mapperFunc = _mapPlanCache.AddOrUpdate(mapRequest);
+            return mapperFunc;
         }
 
         public Delegate GetMapperFunc(MapRequest mapRequest)
         {
-            return _mapPlanCache.GetOrAdd(mapRequest).Typed;
+            return _mapPlanCache.GetOrAdd(mapRequest).Typed ??

```

Comment:
I have to try to get it twice.  Cause the first time might fail because of trying to get the included mappings might not work because of circular references, or recursion.  But after that the second time trying to resolve it will work.